pcdsdevices/interface.py

```diff
@@ -1646,6 +1656,77 @@ def transmission(self):
                 return 1
 
 
+class LightpathMixin(Device):
```

Comment:
I am somewhat out of the loop on this, so grain of salt:

I had assumed the mixin itself would go away and we'd end up with a signal like:

`lightpath = Cpt(LightpathSignal, signals=["sig1", "sig2"], status_func=get_lightpath_status)`

Or possibly just retrieving those from the parent class/instance on `__init__`.

I'm not claiming this is in any way _better_. Before updating all of the device classes it may be useful to discuss the benefits of the current approach over other options with an eye toward future maintainability.

Reply:
I thought the mixin was a simple way to wrap up the functionality, which boils down to adding the `SummarySignal` and requiring the `get_lightpath_status` method.  I thought it was clearer this way, though I will admit that I didn't think much about a signal that sets a method.  (Component internals scare me slightly).  

I felt the main drawback of the old mixin was that the interface had many too many bits to implement (insert, remove, etc).  Is inheriting from the mixin itself a step that users find onerous? 

I guess the other question is: is the proposed API adequate?  (Outside of the implementation details)

This is the kind of feedback I was looking for, happy to consider other implementation approaches 👍

Reply:
> I thought the mixin was a simple way to wrap up the functionality, which boils down to adding the SummarySignal and requiring the get_lightpath_status method. I thought it was clearer this way, though I will admit that I didn't think much about a signal that sets a method. (Component internals scare me slightly).

I think the internals of component - however scary - aren't so necessary to understand. In this case, looking at it as "Component instantiates a `SummarySignal` that has a `.parent` that leads you back to the relevant Device instance (with the important methods)" is probably good enough?

> Is inheriting from the mixin itself a step that users find onerous?

As much as I tend to (over)use it for its code duplication-reducing aspects, I think multiple inheritance should be avoided wherever possible.

> I guess the other question is: is the proposed API adequate? (Outside of the implementation details)

Given the goals, I think it's sufficient. Including the split of `get` vs `calculate` detailed in other conversations below is important, too, I think.

Reply:
(summarizing external discussions)
After some consideration, I think a mixin is appropriate here.  The key point is that we will be storing branch information (`input_branches` and `output_branches`) externally in the happi database, and requiring them as init kwargs.  In doing so, we need to store that information and use it in the calc_lightpath_state methods; this is a pattern that lends itself to mixins.